plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,186 @@
+ @
+ After inlining @original@ we get:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((force^|T1|
+             (delay^|T1|
+               (\X1 ->
+                  delay^|T2| (\X2 ->
+                    delay^|T3| (\X3 ->
+                      ... ->
+                        delay^|Tn| (\Xn -> body)))))
+            ) X1)) X2)) X3 ...)) Xn
+ @
+ In the end, after applying the "UntypedPlutusCore.Transform.ForceDelay" optimisation:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((\X1 ->
+             delay^|T2| (\X2 ->
+               delay^|T3| (\X3 ->
+                 ... ->
+                   delay^|Tn| (\Xn -> body)))))
+        X1) X2)) X3 ...)) Xn
+ @
+ Notice that the next two reduction steps (applying @X1@ and reducing @force (delay ...)@)
+ produce an equivalent term to applying the transformation and then the reduction rule
+ for application.
+ This is easy to check, so we continue by showing what the "optimised term" looks like:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((\X1 -> \X2 ->
+           delay^|T3| (\X3 ->
+             ... ->
+               delay^|Tn| (\Xn -> body))
+        X1) X2)) X3 ...)) Xn
+ @
+ The term can be optimised further by "erasing" the @force^|T3|@ and @delay^|T3|@ pair,
+ and so on until @Tn@.
+-}
+{-# LANGUAGE LambdaCase #-}
+module UntypedPlutusCore.Transform.ForceApply
+    ( forceApply
+    ) where
+
+import UntypedPlutusCore.Core
+
+import Control.Lens (transformOf)
+import Control.Monad (guard)
+import Data.Foldable (foldl')
+import Data.Maybe (fromMaybe)
+
+{- | Traverses the term, for each node applying the optimisation
+ detailed above. For implementation details see 'optimisationProcedure'.
+
+ Note: this optimisation should be applied after
+ 'UntypedPlutusCore.Transform.ForceDelay.forceDelay' for reasons mentioned
+ above.
+-}
+forceApply :: Term name uni fun a -> Term name uni fun a
+forceApply = transformOf termSubterms processTerm
+
+{- | Checks whether the term is of the right form, and "pushes"
+ the 'Force' down into the underlying lambda abstractions.
+-}
+processTerm :: Term name uni fun a -> Term name uni fun a
+processTerm = \case
+    original@(Force _ subTerm) ->
+        fromMaybe original (optimisationProcedure subTerm)
+    t -> t
+
+{- | Converts the subterm of a 'Force' into specialised types for representing
+ multiple applications on top of multiple abstractions. Checks whether the lambda
+ will eventually get "fully reduced" and applies the optimisation.
+ Returns 'Nothing' if the optimisation cannot be applied.
+-}
+optimisationProcedure :: Term name uni fun a -> Maybe (Term name uni fun a)
+optimisationProcedure term = do
+    asMultiApply <- toMultiApply term
+    innerMultiAbs <- toMultiAbs . toApply $ asMultiApply
+    guard $ length (applyToIter asMultiApply) == length (absVars innerMultiAbs)
+    case rhs innerMultiAbs of
+        Delay _ subTerm ->
```

Comment:
So since we have a separate force-delay pass anyway, why do we insist on finding `Delay` here? Why not simply turn

```
force (apply (apply (\x1 x2 -> body) e1) e2)
```

into

```
(apply (apply (\x1 x2 -> force body) e1) e2)
```

and let force-delay clean `force body` up if `body` is a `delay`? This transformation is correct even if `body` isn't a `delay`, right?

I'm not sure if it's helpful to implement things this way, but maybe we can implement an additional optimization that remove an outer `delay` from a binding if that binding is always accompanied by a `force` at all call sites within a program. But even aside from that, not requiring a `Delay` to appear there is simpler, so why not lift the restriction?

We could also generalize that to for example transform

```
force (apply (apply (\x1 -> body) e1) e2)
```

into, erm, this:

```
apply (\x1 -> force (apply body e2)) e1
``` 

This has the advantage that now `body` is directly applied to `e2` instead of appearing under a lambda, so maybe that would help the callsite inliner if we had one for UPLC and not just PIR?

Speaking of which, @zliu41 @michaelpj does the callsite inliner recognize cases where a function is fully applied but is also eta-expanded? E.g. `(\x1 -> f x1) e1 e2` for `f` whose arity equals `2`?

Lots of questions...

Reply:
> So since we have a separate force-delay pass anyway, why do we insist on finding Delay here?

I agree that in general our approach is for a pass to do One Thing and let other passes do their One Thing. So I would indeed be inclined to let `ForceDelay` do the cancellation.

> This transformation is correct even if body isn't a delay, right?

I think so. 

> maybe we can implement an additional optimization that removes an outer delay from a binding if that binding is always accompanied by a force at all call sites within a program.

That would also probably be useful, although a bit fiddly perhaps.

> We could also generalize that...

I think there's a simpler way of looking at this family of optimizations more generally. They're exactly let-floating.

```
force (apply (apply (\x1 -> body) e1) e2)
===
force (apply (let x1 = e1 in body) e2)
--->
let x1 = e1 
in force (apply body e2)
```
which makes it more obvious that you can float the "let" out of all sorts of contexts, not just forces but also applications etc. As usual, you have to worry about effects, but we already figured all this out for the PIR let-floater.

Perhaps it would be better to port the PIR let-floater instead of thinking of this as a special case where we "push in forces"?

> Speaking of which, @zliu41 @michaelpj does the callsite inliner recognize cases where a function is fully applied but is also eta-expanded? E.g. (\x1 -> f x1) e1 e2 for f whose arity equals 2?

I don't think so, no.

Reply:
> I think there's a simpler way of looking at this family of optimizations more generally. They're exactly let-floating.

Oh, that's a very good point which I didn't consider, thanks for the example. I'll look into the PIR let-floater approach.